### ISSUE TITLE
KVStore init vs push

### ISSUE THREAD

**Issue by pluskid** [dmlc/mxnet#365]
Hi, I just found out that the example in [the kvstore tutorial](http://mxnet.readthedocs.org/en/latest/python/tutorial.html#initialization) does not work. The expected behavior:

```
>>> kv = mx.kv.create('local') # create a local kv store.
>>> shape = (2,3)
>>> kv.init(3, mx.nd.ones(shape)*2)
>>> a = mx.nd.zeros(shape)
>>> kv.pull(3, out = a)
>>> print a.asnumpy()
[[ 2.  2.  2.]
 [ 2.  2.  2.]]
```

Instead I get
```
>>> kv.pull(3, out = a)
[12:04:48] ./dmlc-core/include/dmlc/logging.h:208: [12:04:48] src/kvstore/./kvstore_local.h:76: Check failed: it != merge_buf_.end() key 3 has not been pushed
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/chiyuan/work/mxnet/mxnet/python/mxnet/kvstore.py", line 229, in pull
    ctypes.c_int(priority)))
  File "/Users/chiyuan/work/mxnet/mxnet/python/mxnet/base.py", line 72, in check_call
    raise MXNetError(py_str(_LIB.MXGetLastError()))
mxnet.base.MXNetError: [12:04:48] src/kvstore/./kvstore_local.h:76: Check failed: it != merge_buf_.end() key 3 has not been pushed
```

**Comment by tqchen** [comment]
@mli Maybe you want to update the tutorial?

I guess the reason was the push is needed before pull in allreduce mode

**Comment by mli** [comment]
we should not place a must push before pull constraint here, since it is not ture for local-on-kvstore and dist-* will fix it later

**Comment by mli** [comment]
should be fixed by #385

**Comment by pluskid** [comment]
@mli Thanks!
